cNVMe/Driver.cpp

```diff
@@ -0,0 +1,320 @@
+
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
+// OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+############################################################################################
+Driver.cpp - An implementation file for the cNVMe 'Driver'
+*/
+
+#include "Driver.h"
+#include "Identify.h"
+#include "Strings.h"
+#include "Tests.h"
+
+#define ADMIN_QUEUE_SIZE 15	// This is 0 based
+
+using namespace cnvme::tests;
+
+namespace cnvme
+{
+	namespace driver
+	{
+		std::string statusToString(Status s)
+		{
+			if (s == SENT_SUCCESSFULLY)
+			{
+				return "Sent Successfully";
+			}
+			else if (s == NO_MATCHING_SUBMISSION_QUEUE)
+			{
+				return "Didn't have a submission queue with the specified id to send the command to";
+			}
+			else if (s == NO_LINKED_COMPLETION_QUEUE)
+			{
+				return "Didn't have a linked completion queue with for the specified submission queue id for the drive to post completion to";
+			}
+			else if (s == TIMEOUT)
+			{
+				return "The command timed out";
+			}
+			else if (s == BUFFER_NOT_LARGE_ENOUGH)
+			{
+				return "The passed in payload was not large enough";
+			}
+			else if (s == INVALID_DATA_DIRECTION)
+			{
+				return "The data direction given was invalid";
+			}
+			else if (s == INVALID_DATA_LENGTH)
+			{
+				return "The data length was invalid";
+			}
+
+			ASSERT("Status not found in statusToString()");
+			return "Unknown";
+		}
+
+		std::string dataDirectionToString(DataDirection d)
+		{
+			if (d = NO_DATA)
+			{
+				return "No Data";
+			}
+			else if (d == READ)
+			{
+				return "Read";
+			}
+			else if (d == WRITE)
+			{
+				return "Write";
+			}
+			else if (d == BI_DIRECTIONAL)
+			{
+				return "Bi-Directional";
+			}
+
+			ASSERT("DataDirection not found in dataDirectionToString()");
+			return "Unknown";
+		}
+
+		Driver::Driver()
+		{
+			// We have a controller... it is not running.
+			auto controllerRegisters = TheController.getControllerRegisters()->getControllerRegisters();
```

Comment:
Can you add "this->" before "TheController" so that it is more obviously a class variable? Gracias.